### PR TITLE
Implement adding failure image, setting scale type and image loader callback functionality ⚡️

### PR DESCRIPTION
**Pull request by HR**
:shipit:

### CODE REVIEW

**Comment by Piasy** [comment]
Thanks for this PR! Could you please use Fresco's failure image API, and implement a tap-to-retry in both Fresco loader and Glide loader?

**Comment by HR** [comment]
Will surely try. Though I'm not really familiar with Glide.
Will the tap to retry UI be provided by the dev and should there be a default one?

**Comment by Piasy** [comment]
I think we can listen click event on the failure image(for Glide).

**Comment by HR** [comment]
Done! I think that we should add a request callback for on success and fail to allow for custom behaviour.

File: BigImageViewer/src/main/java/com/github/piasy/biv/view/BigImageView.java [hunk 1]
```diff
@@ -133,6 +159,29 @@ public void setOnLongClickListener(OnLongClickListener listener) {
         mImageView.setOnLongClickListener(listener);
     }
 
+    public void setFailureImageScaleType(ImageView.ScaleType scaleType) {
+        mFailureImageScaleType = scaleType;
+    }
+
+    public void setFailureImage(Drawable failureImage) {
+        // Failure image is not set
+        if (failureImage == null) return;
```
Comment (Piasy) [inline BigImageViewer/src/main/java/com/github/piasy/biv/view/BigImageView.java]: Please change into:

``` java
if (failureImage == null) {
    return;
}
```

File: BigImageViewer/src/main/java/com/github/piasy/biv/view/BigImageView.java [hunk 2]
```diff
@@ -347,5 +421,19 @@ public void onAnimationRepeat(Animation animation) {
...
     private void doShowImage(File image) {
         mImageView.setImage(ImageSource.uri(Uri.fromFile(image)));
+        mFailureImageView.setVisibility(GONE);
+        mImageView.setVisibility(VISIBLE);
+    }
+
+    @UiThread
+    private void showFailImage() {
+        // Failure image is not set
+        if (mFailureImageView == null) return;
```
Comment (Piasy) [inline BigImageViewer/src/main/java/com/github/piasy/biv/view/BigImageView.java]: This too.

**Comment by Piasy** [comment]
If no `failureImage` specified, just don't show it. And a callback is great!

**Comment by HR** [comment]
Done! Shall I open a new PR for the callback?

**Comment by Piasy** [comment]
Thanks! I think we can do it together with this PR :)

**Comment by HR** [comment]
Sure 👍

**Comment by HR** [comment]
Done! Do we need an example?

**Comment by Piasy** [comment]
It would be great :) And we can also test it in example.

**Comment by HR** [comment]
Done! Shall I also update the README?

**Comment by HR** [comment]
There should be another callback that is always called when the image has been loaded whether it is from cache or network and you just need the image.

**Comment by HR** [comment]
Think we're ready to merge now 😄

File: README.md [hunk 3]
```diff
@@ -1,18 +1,26 @@
 # BigImageViewer
 
-[ ![Download](https://api.bintray.com/packages/piasy/maven/BigImageViewer/images/download.svg) ](https://bintray.com/piasy/maven/BigImageViewer/_latestVersion)
+[![Download](https://api.bintray.com/packages/piasy/maven/BigImageViewer/images/download.svg)](https://bintray.com/piasy/maven/BigImageViewer/_latestVersion)
```
Comment (Piasy) [inline README.md]: There is a bunch of unnecessary changes in README, please discard it.

File: app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
+/*
```
Comment (Piasy) [inline app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java]: Please change file header.

File: FrescoImageLoader/src/main/java/com/github/piasy/biv/loader/fresco/FrescoImageLoader.java [hunk 5]
```diff
@@ -56,11 +59,12 @@
 public final class FrescoImageLoader implements ImageLoader {
 
     private final Context mAppContext;
+    private final Handler mAppHandler;
     private final DefaultExecutorSupplier mExecutorSupplier;
 
     private FrescoImageLoader(Context appContext) {
         mAppContext = appContext;
-
+        mAppHandler = new Handler(appContext.getMainLooper());
```
Comment (Piasy) [inline FrescoImageLoader/src/main/java/com/github/piasy/biv/loader/fresco/FrescoImageLoader.java]: I've a better idea about thread stuff, I'll clean it up later :)

File: README.md (hunk 3 above)
Comment (HR) [inline README.md]: The getting started part?

Comment (Piasy) [inline README.md]: You can see change log...

File: app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java (hunk 4 above)
Comment (HR) [inline app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java]: To...?

File: README.md (hunk 3 above)
Comment (HR) [inline README.md]: I believe a lot of the restructuring especially the getting started part make it a lot more clearer. Every major project has it.

Comment (HR) [inline README.md]: I believe a lot of the restructuring especially the getting started part make it a lot more clearer. Every major project has it.

File: app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java (hunk 4 above)
Comment (Piasy) [inline app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java]: To the same as other files in this repo.

File: FrescoImageLoader/src/main/java/com/github/piasy/biv/loader/fresco/FrescoImageLoader.java (hunk 5 above)
Comment (HR) [inline FrescoImageLoader/src/main/java/com/github/piasy/biv/loader/fresco/FrescoImageLoader.java]: Can you please do that ASAP as I need this for my project now.

File: README.md (hunk 3 above)
Comment (Piasy) [inline README.md]: What's the point of those changes? And more changes after what I've created screenshots.

screen shot 2017-06-23 at 15 29 21

screen shot 2017-06-23 at 15 29 07
screen shot 2017-06-23 at 15 28 57

File: app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java (hunk 4 above)
Comment (HR) [inline app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java]: Android studio generated this 😂

Comment (Piasy) [inline app/src/main/java/com/github/piasy/biv/example/ImageLoaderCallbackActivity.java]: So please do it manually :)

File: FrescoImageLoader/src/main/java/com/github/piasy/biv/loader/fresco/FrescoImageLoader.java (hunk 5 above)
Comment (Piasy) [inline FrescoImageLoader/src/main/java/com/github/piasy/biv/loader/fresco/FrescoImageLoader.java]: I'll try to cut a release tonight at China time :)

File: README.md (hunk 3 above)
Comment (HR) [inline README.md]: Long lines make it [difficult](https://stackoverflow.com/questions/938105/readme-80-character-limit-or-just-let-the-users-editor-handle-it) to read for the user and so it I'm using the standard 80 character limit.

Comment (Piasy) [inline README.md]: Just use other text editor that will soft wrap lines, e.g. vscode.

Comment (HR) [inline README.md]: It is a standard good practice and allows you to use your favourite editor whether it does soft wrap lines or not.

Comment (HR) [inline README.md]: The `<br />` is to provide some spacing between the gif and getting started section to make it easier to read but can remove it if you like.

**Comment by HR** [comment]
Changed `currentImageFile()` to `getCurrentImageFilePath` since it returns the _path_ of the current image file and not the file itself but shouldn't it return the image i.e.` return mCurrentImageFile;`

**Comment by Piasy** [comment]
I'll merge it and cut a release later tonight :) Thanks for your contribution!

**Comment by HR** [comment]
Great! I've made the changes and added the current image file getter.

**Comment by Piasy** [comment]
I think there is no need for both get path and get file API.

**Comment by HR** [comment]
Then just get file?

**Comment by Piasy** [comment]
Fine :)

**Comment by HR** [comment]
Done 😄

**Comment by Piasy** [comment]
`v1.3.0` is released, check it out :)
